pymodbus/client/serial.py

```diff
@@ -318,7 +322,7 @@ def _wait_for_data(self):
             if available and available != size:
                 more_data = True
                 size = available
-            time.sleep(0.01)
+            time.sleep(self.recv_interval)
```

Comment:
Could the problem be that the preceding line:
```
if self.params.timeout is not None and self.params.timeout:
```
is not true in your test setup, so that you bypass the loop ?
"Condition" should wait for your defined timeout, so changing the time.sleep will only make the loop slower/faster, not really have an effect on data being received or not.

Reply:
I think the logic of this delay is as follows: if serial does not receive data within 0.01s, then the message is considered to have been received.
The message I used to test was 256 bytes and was transmitted all the time. After half of the packets are received, no data is received within 0.01s, resulting in a communication error. Extend the time and the problem disappears. And lower baud rates are more likely to cause problems.
I set timeout to 20s, I think it has nothing to do with this timeout setting.

Reply:
That is not the logic, look at “while condition(start)”, that is intented to loop until timeout.

you can always try to set timeout locally outside the while and e.g. make a debug log in the loop, so you ensure it is actually looping.

Reply:
‘’‘python
        while condition(start):
            available = self._in_waiting()
            if (more_data and not available) or (more_data and available == size):
                break
            if available and available != size:
                more_data = True
                size = available
            time.sleep(0.01)
’‘’
First loop: 
    available = 20, more_data = True, size = available
    time.sleep(0.01)  # If the serial has no data in this 0.01s
Second loop:
    available = 20, (more_data and available == size) = True
    break

At this time, there is still data waiting to be received. However, it will not be read again, resulting in incomplete communication.
I guess it may be related to the system scheduling, causing the serial module to fail to run in a certain time slot. I also tried to replace the test host, but this problem still exists.

Reply:
Good analysis, please change the local variables as requested earlier, then I will approve the PR.

Reply:
And you have a formatting problem (black), you can use ./check_ci.sh to run CI locally.